src/format_rules.ml

```diff
@@ -0,0 +1,82 @@
+open Import
+
+let flag_of_kind : Ml_kind.t -> _ =
+  function
+  | Impl -> "--impl"
+  | Intf -> "--intf"
+
+let add_alias_format sctx loc ~dir ~scope action =
+  let alias_conf =
+    { Dune_file.Alias_conf.name = "fmt"
+    ; deps = []
+    ; action = Some (loc, action)
+    ; locks = []
+    ; package = None
+    ; enabled_if = None
+    ; loc
+    }
+  in
+  Simple_rules.alias sctx ~dir ~scope alias_conf
+
+let sv = String_with_vars.virt_text __POS__
+
+let target_var = String_with_vars.virt_var __POS__ "targets"
+
+let run_rule ~target ~exe ~args ~loc =
+  { Dune_file.Rule.targets = Static [target]
+  ; action = (loc, Run (exe, args))
+  ; mode = Standard
+  ; deps = []
+  ; locks = []
+  ; loc
+  }
+
+let diff file1 file2 =
+  Action.Unexpanded.Diff
+    { optional = false
+    ; mode = Text
+    ; file1
+    ; file2
+    }
+
+let rules_for_ocaml_file loc ocamlformat kind path =
+  let flag = flag_of_kind kind in
+  let path = Path.basename path in
+  let formatted = path ^ ".formatted" in
+  let dep x = String_with_vars.make_macro loc "dep" x in
+  let format_rule =
+    run_rule
+      ~loc
+      ~target:formatted
+      ~exe:(dep @@ Path.to_string ocamlformat)
+      ~args:[sv flag; dep path; sv "-o"; target_var]
+  in
+  let diff_action = diff (dep path) (sv formatted) in
+  (format_rule, diff_action)
+
+let ml_kind_of_extension = function
+  | ".ml" -> Some Ml_kind.Impl
+  | ".mli" -> Some Ml_kind.Intf
+  | _ -> None
+
+let gen_rules sctx ~loc ~dir ~scope ~dir_contents =
+  let add_alias action = add_alias_format sctx loc ~dir ~scope action in
+  let setup_rules (format_rule, diff_action) =
+    let _ : Path.t list =
+      Simple_rules.user_rule sctx ~dir ~scope format_rule
+    in
+    add_alias diff_action
+  in
+  match Bin.which "ocamlformat" with
+  | None ->
+    let msg = "Cannot find ocamlformat, skipping.\n" in
+    add_alias (Echo [sv msg])
+  | Some ocamlformat ->
+    let files = Dir_contents.text_files dir_contents in
```

Comment:
This includes generated files. We really need to look into the file tree for the actual files to reformat.

Reply:
Done. I used `Path.drop_build_context_exn` which works, is that the correct function to use?

Reply:
yep